Refactor Instantiate to support instantiating dataclasses

guillaumegenthial opened the pull request:
<!-- Thank you for sending a PR and taking the time to improve Hydra -->

## Motivation

Address https://github.com/facebookresearch/hydra/issues/1183

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan


## Related Issues and PRs

https://github.com/facebookresearch/hydra/issues/1183

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

omry commented:
Thanks @guillaumegenthial!
It will take a few weeks before I can really review it (It looks good at a quick glance but I will have a lot of feedback).
we can discuss some of my feedback on the chat to make things more efficient.

Don't forget to sign the CLA (see message from the bot above, I can't accept it otherwise).

guillaumegenthial commented:
@omry sure thing, don't hesitate to ping me when you have some time.

To summarize the changes
1. remove kwargs, put all the logic in instantiate
2. handle edge cases carefully (depending on convert, config type, etc.)
3. Add a new ParseMode (not necessary for this PR), that allows the user to avoid the conversion to OmegaConf container and operate on simple python containers (dict + list).

I updated the dataclass test, and I think it now behaves as we would expect. Some cases are counter intuitive (convert = NONE and nesting `dict+dataclass` or `list+dataclass`), as dataclasses instances have no choice but to be absorbed by the top-level OmegaConf container.

omry commented:
Some high level feedback before I do a proper review based on your summary:

> 1. remove `_get_kwargs`, put all the logic in instantiate
Sure, it becomes very long though so it would still be good to break some parts into smaller functions to make it more readable.
(Once the rest of the feedback is incorporated)

> 3. Add mode inheritance (parent modes are passed to the children when recursively instantiating)

I am actually not sure about this one but I am guessing you can still override the mode in the child node?

> 4. Add `ParseMode` (not necessary for this PR), that allows the user to avoid the conversion to OmegaConf container and operate on simple python containers (dict + list).

While I definitely want better support for simple containers, two things:
- This opens up a surface area that will need proper testing (and proper definition of what happens with interpolation and type safety in dict mode).
- It's a user facing change that should be documented.
- I think the parse mode can be determined automatically based on the type of the input config object. (dict or DictConfig).

We should probably split it to a followup PR once the initial changes settles in.

Finally, some of the CI tests are tailing (particularly plugins, which are being instantiated with instantiate(). be sure to take a look and fix the issues.

guillaumegenthial commented:
> I am actually not sure about this one but I am guessing you can still override the mode in the child node?

Yes, children modes override the parent's modes if specified

> We should probably split it to a followup PR once the initial changes settles in.

Definitely agree, I included it here because it's easy to do with the new implementation (it kind of demonstrates its flexibility)

> Finally, some of the CI tests are tailing (particularly plugins, which are being instantiated with instantiate(). be sure to take a look and fix the issues.

I see some "version mismatch" errors, I will have a look at it to understand how instantiate has an impact there.

guillaumegenthial commented:
Ok, I'm pretty sure the failing ray launcher plugin tests are not caused by this PR.
A new version of ray was released 16 hours ago.
The ci instance seems to have both versions co-existing somehow (1.1.0, the new one and 1.0.1.post1, the previous one). Not sure why, maybe just some cache that's not been cleared? I see that the wheels paths are set explicitely in the code to `/tmp/wheels/`, maybe that's the reason...

omry commented:
Yup, unrelated.

guillaumegenthial commented:
Simple comment : `ParseMode` is one way to address https://github.com/facebookresearch/hydra/issues/1200

omry commented:
> Simple comment : `ParseMode` is one way to address #1200

Certainly (assuming that use case is really not depending on type validation, interpolation or any other OmegaConf feature).

omry commented on hydra/utils.py:
```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
+    # TODO: kwargs are used as a config now. If it contains dataclass
+    # objects, they will be interpreted as config nodes in the
+    # instantiation result. We might want to change that behavior.
+    _convert_container_targets_to_strings(kwargs)
+    _convert_container_targets_to_strings(config)
```
Could be a good opportunity to simplify this.
Now that OmegaConf supports objects, we should be able to just convert forward.
instead of converting types to strings (and then locate them again), we can convert everything to types.

guillaumegenthial replied:
I agree.

However it seems we can't really store types in OmegaConf containers.

```python
from omegaconf import OmegaConf

class MyClass:
    pass

OmegaConf.create({"_target_": MyClass})
```

raises

```bash
UnsupportedValueType: Value 'MyClass' is not a supported primitive type
	full_key: _target_
	reference_type=Optional[Dict[Union[str, Enum], Any]]
	object_type=dict
```

guillaumegenthial replied:
actually my bad, I forgot the flag {"allow_objects"} so it works as expected

guillaumegenthial replied:
There is some ambiguity with dataclasses though

```python
@dataclass
class MyClass:
    x: int


cfg = OmegaConf.create({"_target_": MyClass}, flags={"allow_objects": True})
print(cfg._target_)  # {'x': '???'}
```

omry replied:
That again, sounds like another case fox boxing the object.

omry commented on hydra/utils.py:
```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
+    config = _merge(config, kwargs) if kwargs else config
+
+    # Override parent modes from config if specified
+    # Cannot use pop with default because of OmegaConf typed config
```
Not sure I understand the statement.
if DictConfig.pop() behaves differently than native dict pop(), please file an issue with a minimal repro against OmegaConf.

guillaumegenthial replied:
I left a comment on an opened issue https://github.com/omry/omegaconf/issues/425.
I'm not sure this is really a bug, more like some type checking functionality of OmegaConf that forces us to "look before trying"

guillaumegenthial replied:
I opened a separate issue at your request here https://github.com/omry/omegaconf/issues/471

omry commented on hydra/utils.py:
```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
+            if convert == ConvertMode.ALL or (
+                convert == ConvertMode.PARTIAL and config._metadata.object_type is None
+            ):
+                return {
+                    key: instantiate(
+                        value, parse=parse, convert=convert, recursive=recursive
+                    )
+                    for key, value in config.items_ex(resolve=True)
+                }
```
I think dict/list comprehension is no longer elegant once it starts to get this complex.
can you switch to a simple loop populating a `ret = {}` object?

guillaumegenthial replied:
switched all multi-line comprehensions to standard assignments

omry commented on hydra/utils.py:
```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
+
+        if recursive:
+            # If ALL or PARTIAL non structured, instantiate in dict
+            if convert == ConvertMode.ALL or (
+                convert == ConvertMode.PARTIAL and config._metadata.object_type is None
+            ):
```
I don't like how PARTIAL is leaking into this area. somehow this feels off.

omry commented on tests/test_utils.py:
```diff
@@ -979,33 +967,109 @@ def test_convert_params(input_: Any, expected: Any, convert_mode: Any) -> None:
             },
-            SimpleDataClass(a={"foo": 99}, b=[1, 99]),
-            id="simple",
+            (
+                SimpleDataClass(
```
You seems to have dropped the simple case that was there initially.

guillaumegenthial replied:
Now the `expected` test argument is the full instantiated config. It appears like it disappeared but it's still here.

omry reviewed (changes requested):
Initial review pass.
see my comments.

meta:
At this point, switch to piling additional changes as new commits. it will make this easier to review.
We can squash everything before we merge.

guillaumegenthial commented on hydra/_internal/utils.py:
```diff
@@ -613,140 +605,62 @@ def _convert_container_targets_to_strings(d: Any) -> None:
+        return config
+
+    # Manual Override if native mapping type
+    if isinstance(config, collections.abc.MutableMapping):
```
done

guillaumegenthial commented on hydra/_internal/utils.py:
```diff
@@ -613,140 +605,62 @@ def _convert_container_targets_to_strings(d: Any) -> None:
+    NOTE: This is an anti-pattern since we mutate the config and return.
+    This allows us not to do copies of input objects but requires the
+    caller to be aware of the input config mutation during the call.
+    """
```
done

guillaumegenthial replied:
I changed the function behavior.
It mutates and does not return, so we're all good.

guillaumegenthial commented on hydra/_internal/utils.py:
```diff
@@ -613,140 +605,62 @@ def _convert_container_targets_to_strings(d: Any) -> None:
+    if isinstance(config, collections.abc.MutableMapping):
+        if not isinstance(overrides, collections.abc.Mapping):
+            raise TypeError(f"Expected type dict but got {type(overrides)}")
+        config = dict(config.items())  # Shallow copy
```
I removed the copy since this function should be ok mutating stuff. 
As a result, I added a recursive copy in `instantiate`.

guillaumegenthial commented on hydra/_internal/utils.py:
```diff
@@ -613,140 +605,62 @@ def _convert_container_targets_to_strings(d: Any) -> None:
+                config[idx] = _merge(config[idx], item)
+            else:
+                config.append(item)
+        return config
```
I checked OmegaConf behavior.
It overrides lists completely so I removed this to be consistent.
A use case I see is

```python
config = [{"x": 1, "y": 2}]
overrides = [{"x": 2}]
_merge(config, overrides)
assert config == [{"x": 2, "y": 2}]  # Would  be [{"x": 2}] with OmegaConf, the y disappeared.

```

omry replied:
I agree.

guillaumegenthial commented on hydra/_internal/utils.py:
```diff
@@ -613,140 +605,62 @@ def _convert_container_targets_to_strings(d: Any) -> None:
+def _resolve_target(
+    target: Union[str, type, Callable[..., Any]]
+) -> Union[type, Callable[..., Any]]:
+    """Pop target from config and import python type or callable."""
```
done

guillaumegenthial commented on hydra/utils.py:
```diff
@@ -1,39 +1,90 @@
+    TARGET = "_target_"
+    CONVERT = "_convert_"
+    PARSE = "_parse_"
+    RECURSIVE = "_recursive_"
```
done

guillaumegenthial commented on hydra/utils.py:
```diff
@@ -1,39 +1,90 @@
+
+    NONE = "none"  # Keep original config type
+    NATIVE = "native"  # Convert to dict / list if OmegaConf config
+    OMEGACONF = "omegaconf"  # Convert config to OmegaConf container
```
Yes, it should be really easy to remove. I'd just like to remove it once we settle some of the other discussions to be sure the implementation is flexible enough (see below for context)

guillaumegenthial commented on hydra/utils.py:
```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
-        if convert == ConvertMode.PARTIAL:
-            final_kwargs = OmegaConf.to_container(
-                final_kwargs, resolve=True, exclude_structured_configs=True
+    # Finalize config (convert targets to strings, merge with kwargs)
+    # TODO: kwargs are used as a config now. If it contains dataclass
+    # objects, they will be interpreted as config nodes in the
+    # instantiation result. We might want to change that behavior.
```
This is the main friction point.

I'd argue that this would be confusing to treat dataclass instances in overrides as objects and not configs.

1. Instantiate should operate on config-like objects, why would the overrides already have some instantiated objects?
2. DataClass become sources of confusion. If config is `{"obj": {"x": 1, "y": 2}}` and kwargs is `obj=SomeDataClass(x=2)`, what do we expect? `{"obj": {"x": 2, "y": 2}}` (current behavior), or `{"obj": SomeDataClass(x=2)`

Also, (not necessarily a good argument), it would make the implementation more complex. 
The current implementation does
1. merge kwargs fully into config
2. forget about kwargs, only look at config

If we want to treat kwargs as objects, we can't do that because dataclasses instances merged into OmegaConf containers would be converted back to OmegaConf containers. This means that we should keep the kwargs alonside the config during the whole instantiation path.

guillaumegenthial commented:
Thank you for reviewing @omry .
I pushed some changes addressing some of your comments.
Before addressing the others, we need to decide how we should treat kwargs, since it has a potentially big impact on the implementation (see my comment above).

omry commented:
> Thank you for reviewing @omry .
> I pushed some changes addressing some of your comments.
> Before addressing the others, we need to decide how we should treat kwargs, since it has a potentially big impact on the implementation (see my comment above).

Please repeat the comment.
I don't know which comment you are thinking about.
Also, at this point we can have another chat session.

omry commented:
looks like there are some conflicts, can you rebase and fix?

guillaumegenthial commented:
I was referring to your comment

> I think now that we support instantiating dictconfigs, we should treat passthrough dictconfigs as passthrough objects and not as configs.
I guess we can make an exception based on the presence of target but I am not sure it's worth the extra complexity.

My answer was 

> This is the main friction point.
I'd argue that this would be confusing to treat dataclass instances in overrides as objects and not configs.
> 1. Instantiate should operate on config-like objects, why would the overrides already have some instantiated objects?
> 2. DataClass become sources of confusion. If config is `{"obj": {"x": 1, "y": 2}}` and kwargs is `obj=SomeDataClass(x=2)`, what do we expect? `{"obj": {"x": 2, "y": 2}}` (current behavior), or `{"obj": SomeDataClass(x=2)`
> Also, (not necessarily a good argument), it would make the implementation more complex. 

> The current implementation does
> 1. merge kwargs fully into config
> 2. forget about kwargs, only look at config

> If we want to treat kwargs as objects, we can't do that because dataclasses instances merged into OmegaConf containers would be converted back to OmegaConf containers. This means that we should keep the kwargs alonside the config during the whole instantiation path.

And I agree, discussing live on the chat would be more efficient

omry commented:
@guillaumegenthial,
I would like to get some progress on this, would you mind jumping to the chat at some point so we can discuss the remaining issues?

booski360 reviewed (approved):
commit

omry commented:
Taking over. Thanks for working on this.
